File: benchmarks/turso-scale/src/fixture.rs
use std::path::{Path, PathBuf};

use crate::model::CellSpec;

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct FixtureLayout {
    output: PathBuf,
    records: usize,
    repetition: usize,
}

impl FixtureLayout {
    pub fn new(output: PathBuf, records: usize, repetition: usize) -> Self {
        Self {
            output,
            records,
            repetition,
        }
    }

    pub fn source_dir(&self) -> PathBuf {
        self.output
            .join("fixtures")
            .join(self.records.to_string())
            .join(format!("run-{}", self.repetition))
    }

    pub fn source_database(&self) -> PathBuf {
        self.source_dir().join("prolly.db")
    }

    pub fn cell_dir(&self, spec: &CellSpec) -> PathBuf {
        self.output
            .join("cells")
            .join(self.records.to_string())
            .join(format!("run-{}", self.repetition))
            .join(spec.operation.as_str())
            .join(spec.pattern.as_str())
            .join(spec.cache_state.as_str())
    }

    pub fn cell_database(&self, spec: &CellSpec) -> PathBuf {
        self.cell_dir(spec).join("prolly.db")
    }

    pub fn clone_for(&self, spec: &CellSpec) -> Result<(), String> {
        clone_fixture(&self.source_dir(), &self.cell_dir(spec))
    }

    pub fn remove_cell(&self, spec: &CellSpec) -> Result<(), String> {
        remove_generated_dir(&self.output.join("cells"), &self.cell_dir(spec))
    }

    pub fn remove_source(&self) -> Result<(), String> {
        remove_generated_dir(&self.output.join("fixtures"), &self.source_dir())
    }
}

pub fn clone_fixture(source: &Path, destination: &Path) -> Result<(), String> {
    let metadata = std::fs::symlink_metadata(source)
        .map_err(|error| format!("failed to inspect {}: {error}", source.display()))?;
    if metadata.file_type().is_symlink() || !metadata.is_dir() {
        return Err(format!(
            "fixture source must be a real directory: {}",
            source.display()
        ));
    }
    if destination.exists() {
        return Err(format!(
            "fixture destination already exists: {}",
            destination.display()
        ));
    }
    copy_directory(source, destination)
}

fn copy_directory(source: &Path, destination: &Path) -> Result<(), String> {
    std::fs::create_dir_all(destination)
        .map_err(|error| format!("failed to create {}: {error}", destination.display()))?;
    for entry in std::fs::read_dir(source)
        .map_err(|error| format!("failed to read {}: {error}", source.display()))?
    {
        let entry = entry.map_err(|error| format!("failed to read fixture entry: {error}"))?;
        let source_path = entry.path();
        let destination_path = destination.join(entry.file_name());
        let metadata = std::fs::symlink_metadata(&source_path)
            .map_err(|error| format!("failed to inspect {}: {error}", source_path.display()))?;
        if metadata.file_type().is_symlink() {
            return Err(format!(
                "fixture contains a symlink: {}",
                source_path.display()
            ));
        }
        if metadata.is_dir() {
            copy_directory(&source_path, &destination_path)?;
        } else if metadata.is_file() {
            std::fs::copy(&source_path, &destination_path).map_err(|error| {
                format!(
                    "failed to copy {} to {}: {error}",
                    source_path.display(),
                    destination_path.display()
                )
            })?;
        } else {
            return Err(format!(
                "fixture contains an unsupported entry: {}",
                source_path.display()
            ));
        }
    }
    Ok(())
}

pub fn directory_bytes(path: &Path) -> Result<u64, String> {
    if !path.exists() {
        return Ok(0);
    }
    let mut total = 0u64;
    for entry in std::fs::read_dir(path)
        .map_err(|error| format!("failed to read {}: {error}", path.display()))?
    {
        let entry = entry.map_err(|error| format!("failed to read directory entry: {error}"))?;
        let metadata = entry
            .metadata()
            .map_err(|error| format!("failed to inspect {}: {error}", entry.path().display()))?;
        if metadata.is_dir() {
            total = total.saturating_add(directory_bytes(&entry.path())?);
        } else if metadata.is_file() {
            total = total.saturating_add(metadata.len());
        }
    }
    Ok(total)
}

pub fn remove_generated_dir(root: &Path, target: &Path) -> Result<(), String> {
    if target == root || !target.starts_with(root) {
        return Err(format!(
            "refusing to remove path outside generated root: {}",
            target.display()
        ));
    }
    if !target.exists() {
        return Ok(());
    }
    let metadata = std::fs::symlink_metadata(target)
        .map_err(|error| format!("failed to inspect {}: {error}", target.display()))?;
    if metadata.file_type().is_symlink() {
        return Err(format!("refusing to remove symlink: {}", target.display()));
    }
    std::fs::remove_dir_all(target)
        .map_err(|error| format!("failed to remove {}: {error}", target.display()))
}

pub fn database_file_bytes(database: &Path) -> Result<(u64, u64, u64, u64), String> {
    let length = |path: &Path| -> Result<u64, String> {
        match path.metadata() {
            Ok(metadata) => Ok(metadata.len()),
            Err(error) if error.kind() == std::io::ErrorKind::NotFound => Ok(0),
            Err(error) => Err(format!("failed to inspect {}: {error}", path.display())),
        }
    };
    let db = length(database)?;
    let wal = length(&PathBuf::from(format!("{}-wal", database.display())))?;
    let shm = length(&PathBuf::from(format!("{}-shm", database.display())))?;
    Ok((db, wal, shm, db.saturating_add(wal).saturating_add(shm)))
}

File: benchmarks/dolt-prolly-sqlite-compare/fixture.go
package main

import (
	"fmt"
	"io"
	"os"
	"path/filepath"
)

type fixtureLayout struct {
	output     string
	records    int
	repetition int
}

func (l fixtureLayout) sourceDir() string {
	return filepath.Join(l.output, "fixtures", fmt.Sprint(l.records), fmt.Sprintf("run-%d", l.repetition))
}
func (l fixtureLayout) sourceDB() string { return filepath.Join(l.sourceDir(), "prolly.db") }
func (l fixtureLayout) cellDir(spec cellSpec) string {
	return filepath.Join(l.output, "cells", fmt.Sprint(l.records), fmt.Sprintf("run-%d", l.repetition), string(spec.operation), string(spec.pattern), string(spec.cacheState))
}
func (l fixtureLayout) cellDB(spec cellSpec) string {
	return filepath.Join(l.cellDir(spec), "prolly.db")
}

func cloneFixture(source, destination string) error {
	info, err := os.Lstat(source)
	if err != nil {
		return err
	}
	if !info.IsDir() || info.Mode()&os.ModeSymlink != 0 {
		return fmt.Errorf("fixture source must be a real directory: %s", source)
	}
	if _, err := os.Lstat(destination); !os.IsNotExist(err) {
		return fmt.Errorf("fixture destination already exists: %s", destination)
	}
	return copyDirectory(source, destination)
}

func copyDirectory(source, destination string) error {
	if err := os.MkdirAll(destination, 0o755); err != nil {
		return err
	}
	entries, err := os.ReadDir(source)
	if err != nil {
		return err
	}
	for _, entry := range entries {
		sourcePath := filepath.Join(source, entry.Name())
		destinationPath := filepath.Join(destination, entry.Name())
		info, err := os.Lstat(sourcePath)
		if err != nil {
			return err
		}
		if info.Mode()&os.ModeSymlink != 0 {
			return fmt.Errorf("fixture contains symlink: %s", sourcePath)
		}
		if info.IsDir() {
			if err := copyDirectory(sourcePath, destinationPath); err != nil {
				return err
			}
			continue
		}
		if !info.Mode().IsRegular() {
			return fmt.Errorf("fixture contains unsupported entry: %s", sourcePath)
		}
		if err := copyFile(sourcePath, destinationPath, info.Mode().Perm()); err != nil {
			return err
		}
	}
	return nil
}

func copyFile(source, destination string, mode os.FileMode) error {
	in, err := os.Open(source)
	if err != nil {
		return err
	}
	defer in.Close()
	out, err := os.OpenFile(destination, os.O_CREATE|os.O_EXCL|os.O_WRONLY, mode)
	if err != nil {
		return err
	}
	_, copyErr := io.Copy(out, in)
	closeErr := out.Close()
	if copyErr != nil {
		return copyErr
	}
	return closeErr
}

func safeRemove(root, target string) error {
	cleanRoot, err := filepath.Abs(root)
	if err != nil {
		return err
	}
	cleanTarget, err := filepath.Abs(target)
	if err != nil {
		return err
	}
	relative, err := filepath.Rel(cleanRoot, cleanTarget)
	if err != nil || relative == "." || relative == ".." || len(relative) >= 3 && relative[:3] == ".."+string(os.PathSeparator) {
		return fmt.Errorf("refusing to remove path outside generated root: %s", target)
	}
	if info, err := os.Lstat(cleanTarget); err == nil && info.Mode()&os.ModeSymlink != 0 {
		return fmt.Errorf("refusing to remove symlink: %s", target)
	} else if os.IsNotExist(err) {
		return nil
	} else if err != nil {
		return err
	}
	return os.RemoveAll(cleanTarget)
}

func sqliteFileBytes(database string) (dbBytes, walBytes, shmBytes, total uint64, err error) {
	length := func(path string) (uint64, error) {
		info, statErr := os.Stat(path)
		if os.IsNotExist(statErr) {
			return 0, nil
		}
		if statErr != nil {
			return 0, statErr
		}
		return uint64(info.Size()), nil
	}
	if dbBytes, err = length(database); err != nil {
		return
	}
	if walBytes, err = length(database + "-wal"); err != nil {
		return
	}
	if shmBytes, err = length(database + "-shm"); err != nil {
		return
	}
	total = dbBytes + walBytes + shmBytes
	return
}
